_posts/2020-01-22-speed-up-rspec-boot-time.md

```diff
@@ -0,0 +1,146 @@
+author: akshith_yellapragada
+---
+
+# RSpec Boot Up Takes Longer Than Your Webpack? No Longer
+
+## Intro
+
+TDD is a very enjoyable way to code, always getting quick feedback that what you're doing is most likely correct, and able to check for regressions very quickly. The issue then becomes when your Specs take longer to run, or in our case, longer to boot up. 
+
+For example, a standard unit test for one of our Models.
+
+```
+Finished in 1.15 seconds (files took 12.04 seconds to load)
+1 example, 0 failures
+```
+
+Having a 10+ second feedback loop, even for a simple one unit test really made TDD not as efficient.
+It made the TDD process not enjoyable, and also became very easy to break concentration losing even more time.
+
+My team recently made it a major initiative to start writing a lot more tests for everything, so being able to speed the boot process up would really help with getting that started.
+
+The first step is to figure out exactly what is going wrong, before trying to fix it. 
+If you don't know where or what the performance issues are, you can't fix them.
+
+## How to Profile Start Up Speed
+
+We first need to profile and see what's happening before we can fix it, [this](https://mildlyinternet.com/code/profiling-rails-boot-time.html) handy snippet lets you plug in directly to the require that's happens whenever any rails app is booted up including when you run RSpec.
+
+Another option is the [Bumbler gem](https://github.com/nevir/Bumbler), if the snippet doesn't work for you.
+
+The first step was to build to build an empty spec file, so that we can check purely the RSpec loading and run time.
+
+```ruby
+require 'rails_helper'
+
+class Empty
+  def self.test
+    true
+  end
+end
+
+describe Empty do
+  it { expect(Empty.test).to be_truthy }
+end
+```
+
+Now if we run RSpec, we see this large block of everything that's being loaded, and we see how it's adding up into the 10+ seconds that it takes to run a single spec.
+
+![alt text]({{ site.url }}/assets/images/2020-01-22-speed-up-rspec-boot-time/all_gems_loading.png "All gemfiles that are loaded on every spec run.")
+
+Here's what our current situation looks like, and this has been measured by taking the average of 5 runs of the empty spec file from above.
+
+`bundle exec rspec specs/benchmarks/empty_spec.rb`
+`Finished in 0.634441 seconds (files took 13.00 seconds to load)`
+
+There must be a way to either make this better, or cache it.
+
+The first option that popped to mind or going through the Gemfile, and adding `require: false` to the gems that didn't need to be loaded every time, and then specifically require them in the app. That could've worked, but honestly it's a lot of work, and possibly wouldn't bring the speed that much down.
+
+The more efficient option felt like using [Spring](https://github.com/rails/spring) to cache the app's dependencies and preload it, but if that doesn't work then it would be possible to start removing the longest to load ones.
```

Comment:
```suggestion
The more efficient option felt like using [Spring](https://github.com/rails/spring) to cache the app's dependencies and preload it, but if that doesn't work then it would be possible to start removing the longest loading ones.
```